bundles/alpha.model/src/alpha/model/util/AffineFactorizer.xtend

```diff
@@ -0,0 +1,391 @@
+	
+	/** Converts a matrix into an expression. Each column is for one of the output dimensions. */
+	def static matrixToExpression(ISLMatrix matrix, ISLSpace space) {
+		return (0 ..< matrix.nbCols)
+			.map[col | columnToExpression(matrix, space, col)]
+			.mergeExpressions
+	}
+
+	/** Converts a column of a matrix into an expression. */
+	def private static columnToExpression(ISLMatrix matrix, ISLSpace space, int col) {
+		// Start with a new, empty affine expression from the domain of the given space.
+		// This only represents the single output being built.
+		var expression = ISLAff.buildZero(space.copy.domain.toLocalSpace)
+
+		// Populate all of the coefficients for the size parameters.
+		val paramCount = space.dim(ISLDimType.isl_dim_param)
+		for (i : 0 ..< paramCount) {
+			val coefficient = matrix.getElementVal(i, col)
+			expression = expression.setCoefficient(ISLDimType.isl_dim_param, i, coefficient)
+		}
+		
+		// Populate all of the coefficients for the input indexes.
+		val inCount = space.dim(ISLDimType.isl_dim_in)
+		for (i : 0 ..< inCount) {
+			val coefficient = matrix.getElementVal(paramCount + i, col)
+			expression = expression.setCoefficient(ISLDimType.isl_dim_in, i, coefficient)
+		}
+		
+		// If there is one remaining row of the matrix, then it encodes a constant.
+		// In this case, populate the constant with that. Otherwise, the constant is 0.
+		// Note: don't let xtend simplify expression.setConstant(constant)
+		// to expression.constant = constant, as that will likely break things.
+		if (matrix.nbRows > paramCount + inCount) {
+			val constant = matrix.getElementVal(paramCount + inCount, col)
+			expression = expression.setConstant(constant)
+		} else {
+			expression = expression.setConstant(0)
+		}
+		
+		// Convert the single expression into a multi-expression,
+		// and set the dimension name of this output to be the name of the output
+		// from the given space associated with this column.
+		val outputName = space.getDimName(ISLDimType.isl_dim_out, col)
+		return expression
+			.toMultiAff
+			.setDimName(ISLDimType.isl_dim_out, 0, outputName ?: "None")
+	}
+
+	/** Merges a set of affine expressions into a single expression via their flat product. */
+	def static mergeExpressions(ISLMultiAff... expressions) {
+		return expressions.reduce[left, right | left.flatRangeProduct(right)]
+	}
+	
+	////////////////////////////////////////////////////////////
+	// Expression Decomposition
+	////////////////////////////////////////////////////////////
+	
+	/**
+	 * Decomposes an expression using the Hermite decomposition of the matrix
+	 * which represents the given expression. The decomposition is returned as
+	 * a key->value pair, where the key has the same domain as the original
+	 * expression, and the value has the same range as the original expression.
+	 * The range of the key is the same as the domain of the value.
+	 */
+	def static decomposeExpression(ISLMultiAff expression) {
```

Comment:
Could we rename this? “Decompose” by itself is already overloaded a lot. When I first saw this called elsewhere by itself I was trying to figure out “how” it was being decomposed. It’s obvious when you dig and see it calls into hermite. It’d be more obvious I think if the word hermite were in the name somewhere. Maybe just `hermiteDecomposition` like the equivalent version that takes a matrix?

Reply:
Addressed in the latest update. See the commit linked below.
- Renamed the `hermiteDecomposition` function (which is for matrices) to `hermiteMatrixDecomposition`.
- Renamed the `decomposeExpression` function (for expressions) to `hermiteExpressionDecomposition`.

I chose not to have them both named `hermiteDecomposition`, as Java doesn't support function overloading with different return types. I'm not 100% sure I like these names (as it sounds like we're decomposing hermite matrices or hermite expressions). Any ideas on better names?

https://github.com/CSU-CS-Melange/alpha-language/pull/7/commits/df3d8ad784e5b792d4dc7ae09498d7c5bdccbb31